### ISSUE TITLE
[bug] Recipe self.dependencies does not contain any elements

### ISSUE THREAD

**Issue by SeanSnyders** [conan-io/conan#10304]
### Environment Details (include every applicable attribute)
  * Operating System+version: macOS 10.15.7
  * Compiler+version: apple-clang 12
  * Conan version: 1.42.2
  * Python version: 3.9

### Description

As documented in https://docs.conan.io/en/1.42/reference/conanfile/dependencies.html#dependencies-interface one can access dependency information through the `self.dependencies` attribute, but (currently) only in the `validate()` and `generate()` methods.
The repro below indicates that this `self.dependencies` object does not contain any of the build dependencies in these two methods.

### Goal
I am trying to externalise build requirements through profiles, but based off logic in the recipe adapt these requirements in the `build_requirements()` method to be forced into the host context via the `force_host_context=True` parameter, e.g.
```
if cross_buidling:
  self.build_requires(dep, force_host_context=True)
```
And for this I need to know the full version of `dep` that has been provided to the recipe to use that full version in the `self.build_requires()` method.
It seems that (currently) the only place to see these dependencies and their versions is in the the `validate()` and `generate()` methods, so I was going to cache the dependency version in the `validate()` method to reuse in the `build_requirements()` method.

### Steps to reproduce (Include if Applicable)

Use recipe and profile below and call command in terminal:
`conan install . -if=build-debug -pr:h default -pr:h test_deps.profile -pr:b default`

`conanfile.py`
```python
from conans import ConanFile

class DependenciesCheck(ConanFile):

    def validate(self):
        self.output.info(f"Running validate...")
        self.output.info(f"self.dependencies [{len(self.dependencies.items())}]: {self.dependencies}")
        for require, dependency in self.dependencies.items():
            self.output.info("Dependency is direct={}: {}".format(require.direct, dependency.ref))

    def generate(self):
        self.output.info("Running generate...")
        self.output.info(f"self.dependencies [{len(self.dependencies.items())}]: {self.dependencies}")
        for require, dependency in self.dependencies.items():
            self.output.info("Dependency is direct={}: {}".format(require.direct, dependency.ref))

```
`test_deps.profile`
```
[settings]
[options]
[build_requires]
cmake/3.19.4
gtest/1.11.0#a35bd6be75af15855a7c8c861888cd90
benchmark/1.6.0#abc2853305ae1834eb47ccb84a715c81
[env]
```

### Logs (Executed commands with output) (Include/Attach if Applicable)

```
conan install . -if=build-debug -pr:h default -pr:h test_deps.profile -pr:b default
Configuration (profile_host):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=12.0
os=Macos
os_build=Macos
[options]
[build_requires]
*: cmake/3.19.4, gtest/1.11.0, benchmark/1.6.0
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=12.0
os=Macos
os_build=Macos
[options]
[build_requires]
[env]

conanfile.py: Running validate...
conanfile.py: self.dependencies [0]: <conans.model.dependencies.ConanFileDependencies object at 0x10940c4c0>
conanfile.py: Installing package
Requirements
Packages
Build requirements
    benchmark/1.6.0 from 'conancenter' - Cache
    cmake/3.19.4 from 'conancenter' - Cache
    gtest/1.11.0 from 'conancenter' - Cache
    openssl/1.1.1k from 'conancenter' - Cache
Build requirements packages
    benchmark/1.6.0:2f7a9fc6a21fda7cdeaab303f436ee769d8e964c - Cache
    cmake/3.19.4:cf36af68367ac960cc1360ef3ea3c735fa475414 - Cache
    gtest/1.11.0:9e8e90672d7a216fdaac8d6a37a0bbd4a25f56a6 - Cache
    openssl/1.1.1k:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Cache

Installing (downloading, building) binaries...
benchmark/1.6.0: Already installed!
gtest/1.11.0: Already installed!
openssl/1.1.1k: Already installed!
cmake/3.19.4: Already installed!
cmake/3.19.4: Appending PATH environment variable: /Users/ssnyder/Development/TmpConanData/cmake/3.19.4/_/_/package/cf36af68367ac960cc1360ef3ea3c735fa475414/bin
conanfile.py: Applying build-requirement: cmake/3.19.4
conanfile.py: Applying build-requirement: gtest/1.11.0
conanfile.py: Applying build-requirement: benchmark/1.6.0
conanfile.py: Applying build-requirement: openssl/1.1.1k
conanfile.py: Generator txt created conanbuildinfo.txt
conanfile.py: Calling generate()
conanfile.py: Running generate...
conanfile.py: self.dependencies [0]: <conans.model.dependencies.ConanFileDependencies object at 0x10940c4c0>
conanfile.py: Aggregating env generators
conanfile.py: Generated conaninfo.txt
conanfile.py: Generated graphinfo
```

**Comment by memsharded** [comment]
Hi @SeanSnyders 

I have written this unit test, trying to reproduce:

```python
def test_dependencies_visit_build_requires_profile():
    # https://github.com/conan-io/conan/issues/10304
    client = TestClient()
    client.save({"conanfile.py": GenConanfile()})
    client.run("create . cmake/0.1@")
    conanfile = textwrap.dedent("""
        from conans import ConanFile
        class Pkg(ConanFile):

            def generate(self):
                for req, dep in self.dependencies.items():
                    self.output.info("REQUIRE: {}!!!".format(dep.ref))
                dep = self.dependencies.build["cmake"]
                self.output.info("CMAKE: {}!!!".format(dep.ref))
        """)
    client.save({"conanfile.py": conanfile,
                 "profile": "[build_requires]\ncmake/0.1"})
    client.run("install . -pr:b=profile -pr:h=profile --build")  # Use 2 contexts
    assert "conanfile.py: REQUIRE: cmake/0.1!!!" in client.out
    assert "conanfile.py: CMAKE: cmake/0.1!!!" in client.out
```
apparently it is working correctly, and I would say it captures correctly your described case.
There might be some other detail that we are missing. Please have a look to the test, in case there is something different there.

**Comment by SeanSnyders** [comment]
Hi @memsharded 

Is it possible for me to run that unit test with my installed Conan 1.42.2? If so, how?

Did you run the repro I described as is before writing your unit test? If so, what did you get as output?

**Comment by memsharded** [comment]
You were right, I overlooked one detail in your repro that was critical: the usage of ``validate()`` function. Please check the new PR with a potential fix, but also take into account:

- At ``validate()`` time in Conan 1.X, the build-requires are not available yet. Because in 1.X, the "host" graph containing the libraries and executables is computed first, completely, including the packages ``package_ids``. 
- For computing the ``package_ids`` it is necessary to run the ``validate()`` method first, because that can produce an invalid package-id
- Once the final ``package_ids`` are computed, based on the ``--build`` policy, the required ``build_requires`` are computed and expanded.

The PR fixes the caching of ``self.dependencies``, so they are available at ``generate()`` time (which is very important for the correct behavior of generators). Thanks very much for reporting this.

That also means that your goal is not possible, at least in 1.X (all of this has been changed in Conan 2.0, the complete graph, including build-requires is computed first, then the ``validate()``, so in this case, ``validate()`` will have the whole information).

My recommendation for specifying build-requires in profiles would be something like:

- Use ``build-requires`` in profiles only for things that could usually be things installed in the developer machines, and devs might want to use their system installed tool instead the Conan package one. A good example could be cmake.
- Use ``build-requires`` (actually better the more modern ``test_requires``) in recipes to specify test dependencies, like gtest, that are not that typical to be installed at the system level in devs machines.
- To control the versions or revisions that are used, that is achieved using different server repositories. You put the gtest revision you want your builds (devs, CI) etc to use in the "develop" repo, once it is ready for that.

Please let me know if this clarifies a bit the issue and if it helps.

**Comment by SeanSnyders** [comment]
Thanks for the info. There is still a lot of intricacies and unknowns on best practices and recommendations for Conan 1 and 2. This is the type of info that should be front and centre on the website.

Following your recommendations, I still have a question that relates to my goal.
In a recipe, how is/should a test requirement like `gtest` be forced to be in the host context? Following my original post of this issue, that in the `build_requirements()` method, I do the following for cross-compilations such that the dependency (like `gtest` or `benchmark` or similar style of dependency) is compiled in and available in the host otherwise something like the unit tests cannot be compiled and run (e.g. via `node.js` for a `Emscripten` build)

```
if cross_buidling:
  self.build_requires(dep, force_host_context=True)
```

For this scenario I need the full version of `dep` (e.g. `gtest`) to use that `self.build_requires()` method with the `force_host_context=True` parameter.
In my recipe, I might have a default version for this dependency set, but I'd like developers to be able to specify/inject their own version of `gtest` (lets say for using a specific build of the dependency for testing or somesuch). So this can be done via profiles in the `[build_requires]` section, but the logic of the recipe (which can be complex) needs to do the `force_host_context=True` part, thus the version is required in this recipe logic, and thus the full version of the dependency.
Which, if I understand you correctly, will not be available in the recipe for me to use at the correct time (i.e. the `build_requirements()` method) according to Conan v1 flow.
After a bit more explanation from my side, how can this be achieved otherwise in Conan v1?

Then for Conan v2 and the future, if I understand you correctly, the dependency graph will have already been calculated by the time the `validate()` method is called. My goal here is not really about the `vaidate()` method, but to augment the build/test/any requirements (e.g. force a dependency to be in the host context) within the recipe based off logic. Where would I achieve this in Conan v2?

**Comment by memsharded** [comment]
> Thanks for the info. There is still a lot of intricacies and unknowns on best practices and recommendations for Conan 1 and 2. This is the type of info that should be front and centre on the website.

Yes, we know. We have managed to update some of the docs (like the "Create Packages - Gettings Started" or the "Editables"). But the problem with, for example the Getting started of the Consuming Packages, is that recipes of ConanCenter haven't updated yet to some of the modern features, and then, it won't work that easily). We are doing a major re-write of the docs for 2.0 that will fix it, but for the 1.X time, seems challenging to get the time to do it.

> n a recipe, how is/should a test requirement like gtest be forced to be in the host context? Following my original post of this issue, that in the build_requirements() method, I do the following for cross-compilations such that the dependency (like gtest or benchmark or similar style of dependency) is compiled in and available in the host otherwise something like the unit tests cannot be compiled and run (e.g. via node.js for a Emscripten build)

The modern approach should be like:

```python
test_requires = "gtest/1.0"

# or 

def build_requirements(self):
    self.test_requires("gtest/1.0")
```

And then, there is indeed a current limitation (not sure if a feature), ``test_requires`` cannot be overriden from the profile ``[build_requires]``. The rationale is:
- ``tool_requires/build_requires`` are tools, like cmake, compiler, executables, etc. Those can be safely upgraded from the consumer side in many situations. It makes sense for a project to be built with a more modern CMake than the one originally used .
- ``test_requires`` are testing libraries. Their are used in user code, and in the general case it is not straightforward to use a different version than the originally used without breaking due to API changes. So, the version evolve in parallel with the library code that it is testing, but not upgraded independently.

I am curious why it is that important for your developers to change the ``gtest`` version, I haven't seen that case so far.

The possible solution moving forward would be to add ``[test_requires]`` definition in profile files. That might not be easy, and we need to investigate what this mean for Conan 2.0 too, but we could have a look.

**Comment by SeanSnyders** [comment]
What are the semantics of `test_requires()`? Why is it singled out from normal `build_requirements()`? What does Conan do with this extra information of it being "test" requirements.
For consistency (and simplicity to not have edge cases) it would be best to consider the same minimum capabilities for all requirements: they should be upgradeable if required through the same mechanism.

I do not agree with that API changes and versioning is more tightly coupled for "test" requirements vs others.
They are all dependencies. Even compilers and toolchains. And thus, behaviour in using them should be consistent from Conan's perspective. Sure, add more metadata to certain requirements to create a more intelligent Conan, but essentially semver is semver for all dependencies. For example, I might like to test out new features in a compiler that might break certain behaviour in generated code, the same can be said for running complex unit or mocking test code where a new version of a mocking library might change behaviour in how it sets up complex partial class mock states.
This argument is even more pronounced with "newer" toolchains like cross-compiling to emscripten or android compared to, say, more "stable" versions of old-school compilers. Stuff changes a lot and one needs to have the flexibility to experiment easily and safely.

**Coming back to my original goal:** I need to make certain requirements (yes, "test" requirements in this use-case, like `gtest`) available in the host context for cross-compiling situations based off logic in the recipe.
Would this be the reason for having `test_requires()`? Extra meta-data for these specific requirements such that Conan can put them into the host context for me for cross-compilations?

Generally this begs the question: would recipes be able to augment the requirements in a generic way? Conan calculates the dependency graph as per what it was provided as input, then the recipe gets a chance to (possibly) augment that, then the rest of the recipe flow.

**Comment by memsharded** [comment]
> What are the semantics of test_requires()? Why is it singled out from normal build_requirements()? What does Conan do with this extra information of it being "test" requirements.

``test_requires`` is a simplified form of ``self.build_requires(..., force_host_context=True)``, to provide a cleaner transition to 2.0, where the ``force_host_context`` argument will not exists.

> For consistency (and simplicity to not have edge cases) it would be best to consider the same minimum capabilities for all requirements: they should be upgradeable if required through the same mechanism.

I wish it was this way, but for ``build_requires``, when they were introduced due to users requests, one of the functional requirements is that every package could have its own version. So ``zlib`` package could require ``cmake/3.20`` and other package could require ``cmake/3.10``, and every package can build with a different cmake version (or any other "build" tool). While this is not that evident with cmake strong forward compatibility, it is something very necessary for many other tools.

That requirement converted the "build_requires" in **private** requires. Every package can define whatever they want, and that will not propagate either upstream or downstream. The result is that they do not behave as regular ``requires``, that indeed propagate. But that also means, it is not possible, by definition, to override them from downstream. We introduced the ``[build_requires]`` in profiles to alleviate this.

> Coming back to my original goal: I need to make certain requirements (yes, "test" requirements in this use-case, like gtest) available in the host context for cross-compiling situations based off logic in the recipe.
Would this be the reason for having test_requires()? Extra meta-data for these specific requirements such that Conan can put them into the host context for me for cross-compilations?

Yes, you can use ``test_requires``, and the missing piece you are lacking here is the possibility to override them from downstream.

Conan 2.0 has already improved over this, providing the ``visible`` trait, that by default is false, but can be made ``True`` for every require that you want, making ``build_requires/tool_requires`` and ``test_requires`` overridable from downstream.

The other alternative would be to consider that possible ``[test_requires]`` override in the profile, but we need to carefully consider and investigate it, because it might not play nicely with the other features in 2.0.

**Comment by SeanSnyders** [comment]
We are currently (for now) freezing on Conan 1.42.2 as there are continuing breaking changes that are happening with every Conan minor version change and incompatibilities with ConanCenter. We keep on having to implement workarounds in our own situations. The current fluctuating situation with Conan is making it really hard to base any product development on it.

Thus, the `test_requires()` is not possible to be used as it was introduced in Conan 1.43. Also, looking at the code changes for that feature, nothing more that `self.build_requires(..., force_host_context=True)` is being performed, so not much more intelligence there. I still need to perform my own logic check for cross-compilations and put the test requirements into the host context such that the library can be correctly compiled and the tests can actually run. I can thus stick to using the `build_requires()` method.
Any inclination to maybe add this type of intelligence into Conan itself for these types of situations? Is my cross-compilation situation generic enough? It seems that currently `test_requires()` is only aimed at helping v1 -> v2 transition.

Conan v2 is not really a viable option at this moment, right?

Then wrt overriding requirements, I would recommend looking at consistent behaviour and provide easy "external" overriding of requirements (e.g. via profiles and as base through the recipe as well for situations where logic collated together and is build upon using `python_requires()` in derived recipes). This also ties in with my suggestion/question on having access to manipulation of the dependency graph within a recipe after it has been calculated but before commencing the rest of the conan flow. Comments on this?

Overriding requirements externally would facilitate situations where a deep graph of dependencies (especially where each package is developed using similar toolsets and approaches as what typically happens within companies) can be easily tested with new minor/bugfix versions for said requirements without investing enormous time in rolling up each package manually in this dependency graph.

**Comment by memsharded** [comment]
> We keep on having to implement workarounds in our own situations. The current fluctuating situation with Conan is making it really hard to base any product development on it.

Yes, the thing is that ConanCenter contributors love living "at head", and they are always pushing for the latest features and improvements, and using them as soon as they become available. Then, the stability commitment in https://docs.conan.io/en/latest/introduction.html#stable, as explained there, cannot cover ConanCenter, and using ConanCenter recipes implies moving to the latest Conan version.

So what most users do in production when they cannot aggressively update latest Conan is to get stability by storing a copy of the ConanCenter recipes/packages they use in production. This might be achievable with an Artifactory remote repo, but basically any server where you install things from ConanCenter, then upload those same things to your own server, will isolate users from ConanCenter changing, and allow 100% guaranteed stability. I have commented it before, but if I hadn't here, around 70-80% of Conan users do not depend on ConanCenter in production directly, but via some of this patterns. And we believe it is a much robust approach than depending directly on production on a central server, which is a recipe for disaster as we have already seen with other central repos as NPM.

> Any inclination to maybe add this type of intelligence into Conan itself for these types of situations? Is my cross-compilation situation generic enough? It seems that currently test_requires() is only aimed at helping v1 -> v2 transition.

Yes, ``test_requires`` is just an alias to ``build_requires(, force_host_context)``, to be syntax-ready for 2.0. It is not a different feature or functionally different.

> Conan v2 is not really a viable option at this moment, right?

Conan 2.0 is in alpha.2 right now, so no, definitely not viable for production at the moment. Still, it is the reference to have in mind for anything we do in 1.X, because that is the future stable state.

> Then wrt overriding requirements, I would recommend looking at consistent behaviour and provide easy "external" overriding of requirements (e.g. via profiles and as base through the recipe as well for situations where logic collated together and is build upon using python_requires() in derived recipes). This also ties in with my suggestion/question on having access to manipulation of the dependency graph within a recipe after it has been calculated but before commencing the rest of the conan flow. Comments on this?

We wish to implement an easy way for overrides, but we have always hit a problem with anything that is "private" and as such, can diverge. It is very perfectly possible, legal, and even common to have something like one package ``test_requires = "gtest/1.0"`` and a different package ``test_requires = "gtest/2.0"``. Each package source code will be adapted for that specific gtest version, and will not be compiled if the other major version is used.
Now, with this scenario, a normal override is very challenging. If we want to override and use the ``gtest/1.1`` for that package using version 1.0, this override will also override "gtest/2.0". We could invent a syntax like, ``override only the gtest of "pkgA", but do not override the gtest of "pkgB"``, but that syntax will probably be very confusing and problematic at least.

This is the core reason why regular overrides are not an issue, because there is only 1 version of that package in the graph, but doing overrides when there can be more than 1 (and incompatible) version in the graph, it becomes a definition/ux challenge.

We agree to think about this, so as this issue has been closed (the original issue has been addressed by a PR), lets open a new issue from this comment to discuss about possible ``[test_requires]`` definitions in profiles to override upstream defined ones.
